File: Sources/YToolsCore/DebouncedAction.swift
import Foundation

/// Coalesces a burst of main-actor changes into one final operation.
/// Re-scheduling always cancels the previous task, so stale work never runs.
@MainActor
public final class DebouncedAction {
    private var task: Task<Void, Never>?

    public init() {}

    public func schedule(
        after delay: Duration,
        action: @escaping @MainActor @Sendable () -> Void
    ) {
        task?.cancel()
        task = Task { @MainActor [weak self] in
            do {
                try await Task.sleep(for: delay)
            } catch {
                return
            }
            guard !Task.isCancelled else { return }
            self?.task = nil
            action()
        }
    }

    public func cancel() {
        task?.cancel()
        task = nil
    }
}

File: src/YTools.Windows/Core/DebouncedAction.cs
namespace YTools.Core;

/// <summary>
/// Coalesces a burst of changes into one final operation. Re-scheduling always
/// cancels the previous task, so stale work never runs.
/// </summary>
public sealed class DebouncedAction
{
    private CancellationTokenSource? _cancellation;

    public void Schedule(TimeSpan delay, Action action)
    {
        Cancel();
        var cancellation = new CancellationTokenSource();
        _cancellation = cancellation;
        var context = SynchronizationContext.Current;

        _ = Task.Delay(delay, cancellation.Token).ContinueWith(
            completed =>
            {
                if (completed.IsCanceled || cancellation.IsCancellationRequested)
                {
                    return;
                }

                if (context is not null)
                {
                    context.Post(_ => action(), null);
                }
                else
                {
                    action();
                }
            },
            CancellationToken.None,
            TaskContinuationOptions.ExecuteSynchronously,
            TaskScheduler.Default);
    }

    public void Cancel()
    {
        _cancellation?.Cancel();
        _cancellation = null;
    }
}
